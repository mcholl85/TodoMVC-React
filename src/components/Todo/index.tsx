import { useEffect, useRef, useState } from 'react'

type TodoProps = {
  id: string
  title: string
  completed: boolean
  setCompleted: (params: { id: string; completed: boolean }) => void
  setTitle: (params: { id: string; title: string }) => void
  isEditing: string
  setIsEditing: (completed: string) => void
  removeTodo: (id: string) => void
}

export default function Todo({
  id,
  title,
  completed,
  setCompleted,
  setTitle,
  removeTodo,
  isEditing,
  setIsEditing,
}: TodoProps) {
  const [editedTitle, setEditedTitle] = useState(title)
  const inputEditTitle = useRef<HTMLInputElement>(null)

  const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
    if (e.key === 'Enter') {
      setEditedTitle(editedTitle.trim())

      if (editedTitle !== title) {
        setTitle({ id: id, title: editedTitle })
      }
      if (editedTitle === '') removeTodo(id)
      setIsEditing('')
    }
    if (e.key === 'Escape') {
      setEditedTitle(title)
      setIsEditing('')
    }
  }

  useEffect(() => {
    inputEditTitle.current?.focus()
  }, [isEditing])

  return (
    <>
      <div className='view'>
        <input
          className='toggle'
          checked={completed}
          type='checkbox'
          onChange={(e) => setCompleted({ id: id, completed: e.target.checked })}
        />
        <label>{title}</label>
        <button className='destroy' onClick={() => removeTodo(id)}></button>
      </div>
      <input
        className='edit'
        type='text'
        value={editedTitle}
        onChange={(e) => setEditedTitle(e.target.value)}
        onKeyDown={handleKeyDown}
        onBlur={() => setIsEditing('')}
        ref={inputEditTitle}
      />
    </>
  )
}
