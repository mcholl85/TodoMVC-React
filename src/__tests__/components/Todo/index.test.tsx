import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Todo from '../../../components/Todo'

describe('Component Todo', () => {
  test('Matches DOM Snapshot', () => {
    const { asFragment } = render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={() => {}}
        removeTodo={() => {}}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test('Focus the title input, change the list style given a double click on the list', () => {
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={() => {}}
        removeTodo={() => {}}
      />,
    )
    const liEl: HTMLLIElement = screen.getByRole('listitem')

    userEvent.dblClick(liEl)
    expect(screen.getByRole('textbox')).toHaveFocus()
    expect(liEl).toHaveClass('editing')
  })
  test.each`
    value
    ${true}
    ${false}
  `('Change the completed property given a click on checkbox', ({ value }) => {
    const mockSetCompleted = jest.fn()

    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={value}
        setCompleted={mockSetCompleted}
        setTitle={() => {}}
        removeTodo={() => {}}
      />,
    )
    const checkboxEl: HTMLInputElement = screen.getByRole('checkbox')

    userEvent.click(checkboxEl)

    expect(mockSetCompleted).toBeCalledTimes(1)
    expect(mockSetCompleted).toBeCalledWith('1', !value)
  })
  test('Remove the todo given a click on destroy button', () => {
    const mockRemoveTodo = jest.fn()
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={() => {}}
        removeTodo={mockRemoveTodo}
      />,
    )
    const button: HTMLButtonElement = screen.getByRole('button')

    userEvent.click(button)

    expect(mockRemoveTodo).toHaveBeenCalledTimes(1)
  })
  test('edit the todo title after typing value and valid with enter', () => {
    const mockSetTitle = jest.fn()
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={mockSetTitle}
        removeTodo={() => {}}
      />,
    )
    const inputEl = screen.getByRole('textbox')

    userEvent.type(inputEl, 'test{enter}')

    expect(inputEl).toHaveValue('atest')
    expect(mockSetTitle).toHaveBeenCalledWith('1', 'atest')
    expect(mockSetTitle).toHaveBeenCalledTimes(1)
  })
  test('remove the editing todo title after typing value and invalid with espace', () => {
    const mockSetTitle = jest.fn()
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={mockSetTitle}
        removeTodo={() => {}}
      />,
    )
    const inputEl = screen.getByRole('textbox')

    userEvent.type(inputEl, 'test{escape}')

    expect(inputEl).toHaveValue('a')
    expect(mockSetTitle).not.toBeCalled()
  })
  test('cancel the value after typing value then escape', () => {
    const mockSetTitle = jest.fn()
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={mockSetTitle}
        removeTodo={() => {}}
      />,
    )
    const inputEl = screen.getByRole('textbox')

    userEvent.type(inputEl, 'test{escape}')

    expect(inputEl).toHaveValue('a')
    expect(mockSetTitle).not.toHaveBeenCalled()
  })
  test('remove the spaces from the end of the todo title', () => {
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={() => {}}
        removeTodo={() => {}}
      />,
    )
    const inputEl = screen.getByRole('textbox')

    userEvent.type(inputEl, '  {enter}')
    expect(inputEl).toHaveValue('a')

    userEvent.type(inputEl, '  {escape}')
    expect(inputEl).toHaveValue('a')
  })
  test('remove the todo when edited todo has an empty string', () => {
    const mockRemoveTodo = jest.fn()
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={() => {}}
        removeTodo={mockRemoveTodo}
      />,
    )
    const inputEl = screen.getByRole('textbox')

    userEvent.type(inputEl, '{Backspace}{enter}')

    expect(mockRemoveTodo).toHaveBeenCalledTimes(1)
    expect(mockRemoveTodo).toHaveBeenCalledWith('1')
  })
  test('render a list with the class completed when todo is checked', () => {
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={true}
        setCompleted={() => {}}
        setTitle={() => {}}
        removeTodo={() => {}}
      />,
    )

    const liEl: HTMLLIElement = screen.getByRole('listitem')
    const checkboxEl: HTMLInputElement = screen.getByRole('checkbox')

    expect(checkboxEl).toBeChecked()
    expect(liEl).toHaveClass('completed')
  })
  test('do not render a list with the class completed when todo is not checked', () => {
    render(
      <Todo
        id={'1'}
        title={'a'}
        completed={false}
        setCompleted={() => {}}
        setTitle={() => {}}
        removeTodo={() => {}}
      />,
    )

    const liEl: HTMLLIElement = screen.getByRole('listitem')
    const checkboxEl: HTMLInputElement = screen.getByRole('checkbox')

    expect(checkboxEl).not.toBeChecked()
    expect(liEl).not.toHaveClass('completed')
  })
})
