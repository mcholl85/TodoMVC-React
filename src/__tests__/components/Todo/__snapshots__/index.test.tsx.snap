// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Todo Matches DOM Snapshot 1`] = `
<DocumentFragment>
  <li
    class=""
  >
    <div
      class="view"
    >
      <input
        class="toggle"
        type="checkbox"
      />
      <label>
        a
      </label>
      <button
        class="destroy"
      />
    </div>
    <input
      class="edit"
      type="text"
      value="a"
    />
  </li>
</DocumentFragment>
`;
